### PR TITLE
VLM: RasterSourceMetadata collection

### PR DESCRIPTION
**Pull request by pomadchin**
Thie PR adds metadata collection and conversion for the VLM API

### CODE REVIEW

File: vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala [hunk 1]
```diff
@@ -94,6 +94,9 @@ class GeoTiffReprojectRasterSource(
     new GeoTiffReprojectRasterSource(uri, targetCRS, options)
 
   def resample(resampleGrid: ResampleGrid, method: ResampleMethod): RasterSource =
-    new GeoTiffResampleRasterSource(uri, resampleGrid, method)
+    new GeoTiffResampleRasterSource(uri, resampleGrid, method) {
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala]: Thats why i had metadata mismatch. In fact chained operations didn't work. Chained means `RasterSource.reproject.resample... etc`

Comment (echeipesh) [reply vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala]: `GeoTiffReprojectRasterSource!.resample` should produce `GeoTiffReprojectRasterSource` with modified rasterExtent, not `GeoTiffResampleRasterSource`

File: vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      widthCollected shouldBe (widthResampled +- 1e-7)
+      heightCollected shouldBe (heightResampled +- 1e-7)
+
+      summaryCollected.extent shouldBe summaryResampled.extent
+      summaryCollected.cells shouldBe summaryResampled.cells
+      summaryCollected.count shouldBe summaryResampled.count
+    }
+  }
+
+  it("should create ContextRDD from RDD of RasterSources") {
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala]: that's actually what @jbouffard needs

File: vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+      rasterSource.crs shouldBe summary.crs
+      rasterSource.extent shouldBe summary.extent
+      rasterSource.cellSize shouldBe summary.cellSize
+      rasterSource.cellType shouldBe summary.cellType
+      rasterSource.size shouldBe summary.cells
+      files.length shouldBe summary.count
+    }
+
+    // TODO: the problem is in a GDAL -tap parameter usage
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala]: this is the `-tap` parameter usage issue

File: vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+      summaryCollected.cells shouldBe summaryResampled.cells
+      summaryCollected.count shouldBe summaryResampled.count
+    }
+  }
+
+  // TODO: fix JNI management here
+  // this test works as expected, though crashes with a core dump exception
+  // works slower than a GeoTiff version, there is an interesting discussion here: https://github.com/OpenDroneMap/OpenDroneMap/issues/778
+  // also mb we should pass a resampled RasterExtent like we did in a GeoTiff case
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala]: This test can work only after the class path cleanup

File: vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala (hunk 2 above)
Comment (jbouffard) [reply vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala]: :+1:

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import geotrellis.util._
+import geotrellis.vector.{Extent, ProjectedExtent}
+import org.apache.spark.rdd.RDD
+
+case class RasterSummary(
+  crs: CRS,
+  cellType: CellType,
+  cellSize: CellSize,
+  extent: Extent,
+  cells: Long,
```
Comment (jbouffard) [inline vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: This doesn't need to happen for this PR, but based on our discussion earlier, how would you feel about `RasterSummary` having both `GridBounds` and the number of non-`NoData` values in a given `RasterSource`?

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+object RasterSummary {
+  /** Collect [[RasterSummary]] from unstructred rasters, grouped by CRS */
+  def collect[V <: CellGrid: GetComponent[?, ProjectedExtent]](rdd: RDD[V]): Seq[RasterSummary] = {
+    rdd
+      .map { grid =>
+        val ProjectedExtent(extent, crs) = grid.getComponent[ProjectedExtent]
+        val cellSize = CellSize(extent, grid.cols, grid.rows)
+        (crs, RasterSummary(crs, grid.cellType, cellSize, extent, grid.size, 1))
```
Comment (jbouffard) [inline vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: @echeipesh Did we want the number of non-`NoData` values for this PR? Or can that be pushed off until later?

File: vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala [hunk 7]
```diff
@@ -28,15 +28,15 @@ class GeoTiffReprojectRasterSource(
...
   val crs: CRS,
   val options: Reproject.Options = Reproject.Options.DEFAULT
-) extends RasterSource {
-  @transient private lazy val tiff: MultibandGeoTiff =
+) extends RasterSource { self =>
+  @transient lazy val tiff: MultibandGeoTiff =
     GeoTiffReader.readMultiband(getByteReader(uri), streaming = true)
 
-  private lazy val baseCRS = tiff.crs
-  private lazy val baseRasterExtent = tiff.rasterExtent
```
Comment (echeipesh) [inline vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala]: Why remove all the `private` ? Not a problem now but eventually I could see this becoming part of the API unintentionally.

File: vlm/src/main/scala/geotrellis/contrib/vlm/gdal/GDALBaseRasterSource.scala [hunk 8]
```diff
@@ -6,11 +6,14 @@ import geotrellis.raster._
 import geotrellis.raster.reproject.Reproject
 import geotrellis.raster.resample.ResampleMethod
 import geotrellis.vector._
+
 import org.gdal.gdal.{Dataset, gdal}
 import org.gdal.osr.SpatialReference
 
 trait GDALBaseRasterSource extends RasterSource {
+  val uri: String
```
Comment (jbouffard) [inline vlm/src/main/scala/geotrellis/contrib/vlm/gdal/GDALBaseRasterSource.scala]: Why did you add `uri` here as a `val` when `RasterSource` already has it as a method?

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  def toTileLayerMetadata(layoutDefinition: LayoutDefinition, zoom: Int): (TileLayerMetadata[SpatialKey], Some[Int]) =
+    toTileLayerMetadata(LayoutLevel(zoom, layoutDefinition))
+
+  def toTileLayerMetadata(layoutType: LayoutType): (TileLayerMetadata[SpatialKey], Option[Int]) = {
+    val (ld, zoom) = layoutType.layoutDefinitionWithZoom(crs, extent, cellSize)
+    val dataBounds: Bounds[SpatialKey] = KeyBounds(ld.mapTransform.extentToBounds(extent))
+    TileLayerMetadata[SpatialKey](cellType, ld, extent, crs, dataBounds) -> zoom
+  }
+
+  def resample(resampleGrid: ResampleGrid): RasterSummary = {
```
Comment (jbouffard) [inline vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: Should this be moved to `RasterSummarySpec`? Since earlier we talked about how it doesn't make much sense for `RasterSummary` to have a `resample` method.

Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: I think this function should still be there, as it can be very useful for the test purposes, at least for now.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 6 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: At this point it's not resolved actually :/ I'd like to resolve in a separate PR. At the moment cells is just a `cols x rows` value.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 5 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: What's wrong with it? GridBounds are the actual bounds and number of NoData values is a `cells` number. This value is used to estimate better partition sizes as you remember from our conversation. At this point `cells` just computed a bit incorrect, see the comment below.

File: vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala (hunk 7 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/GeoTiffReprojectRasterSource.scala]: Ah will fix it, forgot to revert changes after the debug process.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 5 above)
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: Oh, I misspoke. I meant that it might be useful to have the number of all pixels and the number of non-`NoData` values

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 6 above)
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: That sounds good to me :+1:

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 9 above)
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: That makes sense. I think we can move it to to the tests at some point instead of removing it.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 5 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: Number of all pixels would be GridBounds size.

Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: Ah, that's true.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala (hunk 9 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSummary.scala]: @jbouffard that's true I added a TODO thing, probably i will remove it in a separate PR!

File: vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala (hunk 4 above)
Comment (pomadchin) [reply vlm/src/test/scala/geotrellis/contrib/vlm/RasterSummarySpec.scala]: the dump contains the following: 

```
# Problematic frame:
# C  [libgdal.20.dylib+0x4a430e]  GDALOverviewBand::FlushCache()+0x14
```
